Таратухин Сергей

taratuhin opened the pull request:
Добавлен проект
Добавлен файл CHANGELOG.md

taratuhin committed:
Добавлен проект
Добавлен файл CHANGELOG.md

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,24 @@
+            onTriggered: {
+                var component = Qt.createComponent("button.qml")
+                var button = component.createObject(mainWindow)
+                button.show()
```
Интервал таймера создания нужно менять при каждой сработке случайно, см. ТЗ.

Toxin65 commented on button.qml:
```diff
@@ -0,0 +1,52 @@
+        id: btn
+        width: 20
+        height: 20
+        x: Math.random() * 280// + width
```
Надо у родителя брать. Кажется `parent.width` должен работать.
И делать это надо не биндингом тогда уж.

И ниже такое же замечание.

Toxin65 commented on button.qml:
```diff
@@ -0,0 +1,52 @@
+            hoverEnabled: true
+            property bool hovered: false
+
+            onEntered: hovered = true
```
А кто будет обратно в false возвращать?
Я думаю `MouseArea` излишне.

Toxin65 commented on button.qml:
```diff
@@ -0,0 +1,52 @@
+            if (btn.y > mainWindow.height)
+            {
+                mainWindow:title = qsTr("Game Over!")
+                mainWindow:color = "Red"
```
Обращение через двоеточие? Это интересно, никогда такого не видел.

Toxin65 reviewed (changes requested):
Есть несоответствие ТЗ.

taratuhin committed:
Изменён интервал появления объектов, объекты появляются случайным образом в интервале от 0.1 до 1 секунды.
Изменена область появления объектов, объекты появляются по всей ширине окна с учётом изменения ширины окна.
Упрощён механизм определения наличия курсора мыши над объектом.

taratuhin committed:
Исправлена ошибка при формировании интервала появления объектов.

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,25 @@
+            onTriggered: {
+                var component = Qt.createComponent("button.qml")
+                var button = component.createObject(mainWindow)
+                button.show
```
А скобочки не нужны разве?

taratuhin replied:
Если поставить скобочки приложение запускается, но в консоль периодически вылазит сообщение:
TypeError: Property 'show' of object QQuickItem(0x5baed0) is not a function

Toxin65 replied:
Значит что-то не так! Это выражение пустое, оно ни к чему не обязывает вообще.

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,25 @@
+            repeat: true
+            running: true
+            onTriggered: {
+                var component = Qt.createComponent("button.qml")
```
В условии не было, но зачем было разбивать на два QML непонятно.

Toxin65 commented on button.qml:
```diff
@@ -0,0 +1,45 @@
+            }
+            else
+            {
+                btn.y += 0.25
```
Ого, а почему скорости у всех одинаковые, по ТЗ скорости падения разные от кнопки к кнопке.

taratuhin replied:
Скорость у всех разная. Для каждой кнопки создается свой таймер с разным временем срабатыванием, а это приращение перемещения.

Toxin65 replied:
Это не очень хороший подход. Но имеет право на жизнь. Хотя я игрался и индивидуальной скоростью и таймером.
Лучший вариант, кмк, который предложили в этом году, общий таймер перемещения для всех, а скорости у всех свои.

taratuhin replied:
Пока не понимаю как событие срабатывания одного таймера связать со всеми динамически созданными объектами, чтобы у каждого объекта в Component можно было увеличить координату `y` на случайное число для каждого объекта и при этом чтоб объекты двигались плавно.

Toxin65 commented on button.qml:
```diff
@@ -0,0 +1,45 @@
+import QtQuick.Controls
+
+
+Item {
```
Этот `Item` у вас вместо главного окна, по сути. Зачем?

Toxin65 commented on button.qml:
```diff
@@ -0,0 +1,45 @@
+Item {
+    id: itm
+    width: mainWindow.width
+    height: mainWindow.height
```
Ой как некрасиво! Куча `Item` ов друг на друге будет, нееее, так совсем нехорошо. Вертайте в один файл всё и через `Component`

Toxin65 reviewed (changes requested):
Похоже опять есть нарушение ТЗ.

taratuhin committed:
Удалил файл button.qml
Объекты создаются через createQmlObject().

Toxin65 commented on Main.qml:
```diff
@@ -15,9 +15,45 @@ Window {
-                var component = Qt.createComponent("button.qml")
-                var button = component.createObject(mainWindow)
-                button.show
+                var component = Qt.createQmlObject('
```
Да что же ты делаешь!!!
Вынеси в `Component` уже!!! )))

taratuhin committed:
Изменено создание объектов, объекты создаются через Component.

taratuhin committed:
Удалён не нужный код.

taratuhin committed:
Удалена пустая строка.

taratuhin committed:
Изменён алгоритм движения объектов.

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,67 @@
+    visible: true
+    title: qsTr("QmlMovement")
+
+    Column {
```
А зачем он нужен?

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,67 @@
+
+    Column {
+        Timer {
+            id: m_timer
```
В QML нет таких правил именования, как в плюсах.

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,67 @@
+            id: btn
+            width: 20
+            height: 20
+            x: Math.random() * (mainWindow.width - btn.width)
```
Этот биндинг, думаю. плохо себя ведёт до разрушения. Я бы не стал делать биндинги для координат кнопки вообще!
Внизу задать координаты в Component.onCompleted. Примеры я вам давал.

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,67 @@
+            hoverEnabled: true
+            onClicked: destroy()
+
+            property double dy: Math.random()
```
Тут ноль возможен...
Да и диапазон не определён.

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,67 @@
+                }
+                else
+                {
+                    btn.y += dy
```
Дык запаковать тернарником в одну строку, чё размазывать...

Toxin65 reviewed (changes requested):
В целом неплохо, но есть огрехи, которые надо исправить.

taratuhin committed:
Сделан рефакторинг.
Удалён не нужный код.

taratuhin committed:
Изменён механизм движения объектов (используются сигналы). Все объекты движутся от одного таймера. Таймер формирует сигнал, а объекты его принимают.

Toxin65 commented on Main.qml:
```diff
@@ -19,6 +19,17 @@ Window {
+        running: true
+        signal signalBtnTimer
+        onTriggered: {
+            btnTimer.signalBtnTimer()
```
В качестве примера, конечно неплохо. Но у таймера и так есть сигнал `triggered`, на который и подписан этот обработчик, который испускает уже другой сигнал. Смысла в промежуточном сигнале нет.

taratuhin replied:
Если переделываю на `triggered` вылетает такая ошибка: `QQmlApplicationEngine failed to load component
qrc:/qt/qml/QmlMovement/Main.qml:16:9: Cannot assign to non-existent property "triggered"` Пока не понимаю что происходит.

Toxin65 replied:
Вы не поняли меня, видимо.

Toxin65 commented on Main.qml:
```diff
@@ -31,31 +42,24 @@ Window {
 
-            function btnMove() {
-                btn.hovered ? btn.y += dy * 2 : btn.y += dy
+            Connections {
```
Да, это подходящий способ. Можно ещё динамически.

Toxin65 commented on Main.qml:
```diff
@@ -0,0 +1,65 @@
+
+            Connections {
+                target: btnTimer
+                function onSignalBtnTimer() {
```
Вот эта функция `onTriggered` без левого сигнала.
